File: CMPS3500_MiniScheme_Group07/oop/Tokenizer.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public final class Tokenizer {

    private Tokenizer() {}

    public static TokenStream Tokenize(String _inFilePath) throws IOException {

        TokenStream stream = new TokenStream();
        String content = Files.readString(Path.of(_inFilePath));
        content = content.replace("(", " ( ")
                         .replace(")", " ) ");
        String[] tokens = content.trim().split("\\s+");

        for (String token : tokens) {

            switch (token) {

                case "("      -> stream.enqueue(new Token(TokenType.L_PAREN, token));
                case ")"      -> stream.enqueue(new Token(TokenType.R_PAREN, token));

                case "#t"     -> stream.enqueue(new Token(TokenType.BOOL_TRUE, token));
                case "#f"     -> stream.enqueue(new Token(TokenType.BOOL_FALSE, token));

                case "+"      -> stream.enqueue(new Token(TokenType.OPERATOR_PLUS, token));
                case "-"      -> stream.enqueue(new Token(TokenType.OPERATOR_MINUS, token));
                case "*"      -> stream.enqueue(new Token(TokenType.OPERATOR_MULTIPLY, token));
                case "/"      -> stream.enqueue(new Token(TokenType.OPERATOR_DIVIDE, token));
                case "="      -> stream.enqueue(new Token(TokenType.OPERATOR_EQUALS, token));
                case "<"      -> stream.enqueue(new Token(TokenType.OPERATOR_LESS, token));
                case ">"      -> stream.enqueue(new Token(TokenType.OPERATOR_GREATER, token));
                case "<="     -> stream.enqueue(new Token(TokenType.OPERATOR_LESSEQUALS, token));
                case ">="     -> stream.enqueue(new Token(TokenType.OPERATOR_GREATEREQUALS, token));

                case "if"     -> stream.enqueue(new Token(TokenType.SPECIAL_IF, token));
                case "cond"   -> stream.enqueue(new Token(TokenType.SPECIAL_COND, token));
                case "let"    -> stream.enqueue(new Token(TokenType.SPECIAL_LET, token));
                case "lambda" -> stream.enqueue(new Token(TokenType.SPECIAL_LAMBDA, token));
                case "define" -> stream.enqueue(new Token(TokenType.SPECIAL_DEFINE, token));

                default       -> stream.enqueue(new Token(
                    (Character.isDigit(token.charAt(0)) ? TokenType.INTEGER : TokenType.IDENTIFIER), token));
            }
        }

        return stream;
    }
}

File: CMPS3500_MiniScheme_Group07/procedural/tokenizer.cpp
/**************************************
 * NAME: Adam Gregory
 * FILE: tokenizer.cpp
 * ASGT: CMPS 3500 Group Project
 * DATE: 3/31/2026
 **************************************/

#include <iostream>
#include <string>
#include "helpers.h"
#include <sstream>
#include <vector>
#include <algorithm>
#include <cctype>

void tokenizer(std::string str, 
    std::vector<TokenType>& tokens, 
    std::vector<std::string>& values) {
    //put spaces between parenthesis
    std::string expr;
    for (char c : str) {
        if (c == '(' || c == ')') {
            expr += ' ';
            expr +=   c;
            expr += ' ';
        } else {
            expr +=   c;
        }
    }

    //set up for reading expression
    std::istringstream iss(expr);
    std::string atom;

    //while the stream reads atoms in the expression, push value and type to stack
    while (iss >> atom) {
        //add token value
        values.push_back(atom);

        //add token type
        if (atom == "(") tokens.push_back(L_PAREN);
        else if (atom == ")") tokens.push_back(R_PAREN);

        else if (atom == "#f") tokens.push_back(BOOL_FALSE);
        else if (atom == "#t") tokens.push_back(BOOL_TRUE);

        else if (atom == "+") tokens.push_back(OPERATOR_PLUS);
        else if (atom == "-") tokens.push_back(OPERATOR_MINUS);
        else if (atom == "*") tokens.push_back(OPERATOR_MULTIPLY);
        else if (atom == "/") tokens.push_back(OPERATOR_DIVIDE);
        else if (atom == "=") tokens.push_back(OPERATOR_EQUALS);
        else if (atom == "<") tokens.push_back(OPERATOR_LESS);
        else if (atom == ">") tokens.push_back(OPERATOR_GREATER);
        else if (atom == "<=") tokens.push_back(OPERATOR_LESSEQUALS);
        else if (atom == ">=") tokens.push_back(OPERATOR_GREATEREQUALS);

        else if (atom == "if") tokens.push_back(SPECIAL_IF);
        else if (atom == "cond") tokens.push_back(SPECIAL_COND);
        else if (atom == "let") tokens.push_back(SPECIAL_LET);
        else if (atom == "lambda") tokens.push_back(SPECIAL_LAMBDA);
        else if (atom == "define") tokens.push_back(SPECIAL_DEFINE);

        else if (std::isdigit(atom[0]) || 
                (atom[0] == '-' && std::isdigit(atom[1])))
            tokens.push_back(INTEGER);
        else tokens.push_back(IDENTIFIER);
    }
}